### PR TITLE
`can_replace()` tests

### PR DESCRIPTION
**Pull request by arvidn**
This patch factors out the `can_replace()` function from the mempool class into a free function. It doesn't need to access anything important from the class, and making it a free function makes it easier to unit test.

There are a few more refactors of the function to simplify its interface, to further simplify testing of it. Finally, a unit test is added to cover the interesting cases of this function's behavior.

This is preparing for extending the replace-by-fee rules, for the new `ASSERT_BEFORE_*` conditions.

These changes should be reviewed one commit at a time. Here's a summary of the changes in each commit:

## min_fee_increase is a constant
Remove it as a member function if `Mempool` as it just always returns a constant. Simplify the code by just making it a constant, not a member of the `Mempool` class

## move can_replace() out of the `Mempool` class
It doesn't need access to any information from the class, so it should be a free function. This makes it a lot simpler to unit test

## simplify can_replace() parameters
Instead of passing in `fees` and `fees_per_cost` as separate parameters, pass in the new `MempoolItem` (which includes both of those). This is not an obvious improvement, but it's a preparation for adding logic (in a separate PR) that needs more fields from the new `MempoolItem`

## simplify the `conflicts` parameter
Instead of passing in a dictionary, mapping bundle name to `MempoolItem`, pass in a set of `MempoolItems`. Right now, we build the dictionary outside the call to `can_replace()` just to pull out the `.values()` inside the call. This step also needs to make the mempool item support being placed in a set.

## simplify the `removals` parameter
Instead of passing in a dictionary, mapping coin names to `CoinRecord`, just pass in a list of the coin names being removed. That's all we need, and it greatly simplifies unit tests, to not have to create dummy `CoinRecord` objects. We were already creating the container of `removal_names` (I just changed it to be a set instead of a list).

## add unit tests
Lastly, add a unit test with all interesting edge cases. (Please make sure I didn't miss any).

### CODE REVIEW

File: chia/full_node/mempool_manager.py [hunk 1]
```diff
@@ -439,8 +397,8 @@ async def validate_spend_bundle(
 
         assert npc_result.conds is not None
         # build removal list
```
Comment (AmineKhaldi) [inline chia/full_node/mempool_manager.py:399]: list => set

File: chia/full_node/mempool_manager.py [hunk 2]
```diff
@@ -556,32 +509,32 @@ async def validate_spend_bundle(
             f"Cost: {cost} ({round(100.0 * cost/self.constants.MAX_BLOCK_COST_CLVM, 3)}% of max block cost)",
         )
 
-        return None, potential, list(conflicting_pool_items.keys())
+        return None, potential, [item.name for item in conflicts]
 
-    def check_removals(self, removals: Dict[bytes32, CoinRecord]) -> Tuple[Optional[Err], List[Coin]]:
+    def check_removals(self, removals: Dict[bytes32, CoinRecord]) -> Tuple[Optional[Err], Set[MempoolItem]]:
```
Comment (AmineKhaldi) [inline chia/full_node/mempool_manager.py:514]: I would update the description below to reflect that we're returning a set of mempool items now.

File: tests/core/mempool/test_mempool_manager.py [hunk 3]
```diff
@@ -496,6 +500,74 @@ async def test_ephemeral_timelock(
...
+        ([mk_item(coins[0:1])], mk_item(coins[0:1], fee=9999999), False),
+        # this is the minimum fee increase
+        ([mk_item(coins[0:1])], mk_item(coins[0:1], fee=10000000), True),
+        # FEE RATE RULE
+        # the new item must pay a higher fee per cost than the existing item(s)
+        # the existing fee rate is 2 and the new fee rate is 2
+        ([mk_item(coins[0:1], cost=1000, fee=2000)], mk_item(coins[0:1], cost=10000000, fee=20000000), False),
+        # the new rate is >2
+        ([mk_item(coins[0:1], cost=1000, fee=2000)], mk_item(coins[0:1], cost=10000000, fee=20000001), True),
+        # SUPER SET RULE
```
Comment (AmineKhaldi) [inline tests/core/mempool/test_mempool_manager.py:539]: SUPER SET => SUPERSET

File: tests/core/mempool/test_mempool_manager.py [hunk 4]
```diff
@@ -496,6 +500,74 @@ async def test_ephemeral_timelock(
...
+        # we can't replace an item spending coin 0 and 1 with an
+        # item that just spend 0
```
Comment (AmineKhaldi) [inline tests/core/mempool/test_mempool_manager.py:541]: spend => spends coin

File: tests/core/mempool/test_mempool_manager.py [hunk 5]
```diff
@@ -496,6 +500,74 @@ async def test_ephemeral_timelock(
...
+        ([mk_item(coins[0:2])], mk_item(coins[0:1], fee=10000000), False),
+        # or just spends coin 1
+        ([mk_item(coins[0:2])], mk_item(coins[1:2], fee=10000000), False),
+        # but if we spend the same coins
+        ([mk_item(coins[0:2])], mk_item(coins[0:2], fee=10000000), True),
+        # or if we spend the same coins with additioal coins
```
Comment (AmineKhaldi) [inline tests/core/mempool/test_mempool_manager.py:547]: additional

File: tests/core/mempool/test_mempool_manager.py [hunk 6]
```diff
@@ -496,6 +500,74 @@ async def test_ephemeral_timelock(
...
+        ([mk_item(coins[0:2])], mk_item(coins[0:3], fee=10000000), True),
+        # FEE- AND FEE RATE RULES
+        # if we're replacing two items, each paying a fee of 100, we need to
+        # spend (at least) the same coins and pay at least 10000000 higher fee
+        (
+            [mk_item(coins[0:1], fee=100, cost=100), mk_item(coins[1:2], fee=100, cost=100)],
+            mk_item(coins[0:2], fee=10000200, cost=200),
+            True,
+        ),
+        # if the fee rate exactly the same, we won't allow the replacement
```
Comment (AmineKhaldi) [inline tests/core/mempool/test_mempool_manager.py:557]: exactly => is exactly

File: chia/full_node/mempool_manager.py [hunk 7]
```diff
@@ -556,32 +509,32 @@ async def validate_spend_bundle(
...
         """
         This function checks for double spends, unknown spends and conflicting transactions in mempool.
-        Returns Error (if any), dictionary of Unspents, list of coins with conflict errors (if any any).
+        Returns Error (if any), the sett of existing MempoolItems with conflicting spends (if any any).
```
Comment (AmineKhaldi) [inline chia/full_node/mempool_manager.py:517]: `Returns Error (if any), the set of existing mempool items with conflicting spends (if any).` ?

**Review by AmineKhaldi — DISMISSED** [review]

**Review by AmineKhaldi — APPROVED** [review]

**Review by trepca — APPROVED** [review]

**Pull request merged**